File: next/dweb-browser/src/Platforms/iOS/MicroModule/Browser/Mwebview/MultiWebViewController.cs
using System.Text.Json;
using UIKit;
using WebKit;
using CoreGraphics;

#nullable enable

namespace DwebBrowser.MicroService.Browser.Mwebview;

public partial class MultiWebViewController
{
    static readonly Debugger Console = new("MultiWebViewController");

    public WindowController Win { get; set; }
    public Ipc Ipc { get; init; }
    public MicroModule LocaleMM { get; set; }
    public MicroModule RemoteMM { get; set; }

    public MultiWebViewController(WindowController win, Ipc ipc, MicroModule localeMM, MicroModule remoteMM)
    {
        Win = win;
        Ipc = ipc;
        LocaleMM = localeMM;
        RemoteMM = remoteMM;

        var gesture = new UIScreenEdgePanGestureRecognizer();
        gesture.AddTarget(() => OnScreenEdgePan(gesture));
        gesture.Edges = UIRectEdge.Left;
        gesture.CancelsTouchesInView = true;
        EdgeView.AddGestureRecognizer(gesture);

        WebViewList.OnChangeAdd(async (_, _) =>
        {
            await UpdateStateHook();
        });

        WindowAdapterManager.Instance.RenderProviders.TryAdd(win.Id, Render);

        /// 窗口销毁的时候
        Win.OnClose.OnListener += async (_, _) =>
        {
            /// 移除渲染适配器
            WindowAdapterManager.Instance.RenderProviders.Remove(win.Id, out _);

            /// 清除释放所有的 webview
            foreach (var item in WebViewList)
            {
                await CloseWebViewAsync(item.webviewId);
            }
        };

        /// ipc 断开的时候，强制关闭窗口
        ipc.OnClose += async (_) =>
        {
            await Win.Close(true);
        };
    }

    UIGestureRecognizerState preState = UIGestureRecognizerState.Ended;
    async void OnScreenEdgePan(UIScreenEdgePanGestureRecognizer pan)
    {
        //pan.ShouldRecognizeSimultaneously = new UIGesturesProbe();
        if (pan.State != preState)
        {
            preState = pan.State;
            Console.Log("OnScreenEdgePan", "state: {0}", pan.State);
        }

        if (pan.State == UIGestureRecognizerState.Ended)
        {
            var LastViewOrNull = WebViewList.LastOrNull();
            if (LastViewOrNull is not null)
            {
                if (LastViewOrNull.webView.CanGoBack)
                {
                    LastViewOrNull.webView.GoBack();
                }
                else
                {
                    await CloseWebViewAsync(LastViewOrNull.webviewId);
                }
            }
        }
    }

    public readonly ChangeableList<ViewItem> WebViewList = new();

    internal record ViewItemState(string webviewId, bool isActivated, string url);

    public async Task UpdateStateHook()
    {
        var currentState = new Dictionary<string, ViewItemState>();
        WebViewList.ForEach(it =>
        {
            var viewItemState = new ViewItemState(it.webviewId, WebviewContainer.Hidden, it.webView.Url?.AbsoluteString ?? "");
            currentState.Add(it.webviewId, viewItemState);
        });

        await Ipc.PostMessageAsync(IpcEvent.FromUtf8(EIpcEvent.State.Event, JsonSerializer.Serialize(currentState)));
    }

    private static int s_webviewId_acc = 0;

    private readonly Dictionary<Mmid, Ipc> _mIpcMap = new();

    public record ViewItem(
        string webviewId,
        DWebView.DWebView webView,
        MultiWebViewController mwebviewController,
        WindowController win);

    public async Task<ViewItem> OpenWebViewAsync(string url, WKWebViewConfiguration? configuration = null)
    {
        var dwebview = await CreateDwebView(url, configuration);

        var viewItem = await AppendWebViewAsItemAsync(dwebview);
        Console.Log("openWebView", viewItem.webviewId);

        /// 提供窗口相关的行为
        dwebview.OnCreateWebView += async (args, _) =>
        {
            var item = await OpenWebViewAsync(args.navigationAction.Request.Url.AbsoluteString!, args.configuration);
            args.completionHandler(item.webView);
        };
        dwebview.OnClose += async (args, _) =>
        {
            await CloseWebViewAsync(viewItem.webviewId);
        };

        return viewItem;
    }

    public Task<DWebView.DWebView> CreateDwebView(string url, WKWebViewConfiguration? configuration = null)
    {
        return MainThread.InvokeOnMainThreadAsync(() =>
        {
            var dwebview = new DWebView.DWebView(null, LocaleMM, RemoteMM, new(url), configuration);
            return dwebview;
        });
    }

    public Task<ViewItem> AppendWebViewAsItemAsync(DWebView.DWebView dwebview)
    {
        var webviewId = "#w" + Interlocked.Increment(ref s_webviewId_acc);

        return MainThread.InvokeOnMainThreadAsync(() => new ViewItem(webviewId, dwebview, this, Win).Also(it =>
        {
            WebViewList.Add(it);
            dwebview.OnClose += async (_, _) =>
            {
                await CloseWebViewAsync(webviewId);
            };

            _ = Task.Run(async () =>
            {
                await _WebViewOpenSignal.Emit(webviewId).ForAwait();
            }).NoThrow();
        }));
    }

    public async Task<bool> CloseWebViewAsync(string webviewId)
    {
        var viewItem = WebViewList.Find(viewItem => viewItem.webviewId == webviewId);

        if (viewItem is not null)
        {
            if (WebViewList.Remove(viewItem))
            {
                await MainThread.InvokeOnMainThreadAsync(async () =>
                {
                    viewItem.webView?.Dispose();
                });
                await _WebViewCloseSignal.Emit(webviewId).ForAwait();

                return true;
            }
        }

        return false;
    }

    /// <summary>
    /// 移除webview所有列表
    /// </summary>
    public bool DestroyWebView()
    {
        MainThread.InvokeOnMainThreadAsync(() =>
        {
            WebViewList.ForEach(viewItem =>
            {
                viewItem.webView?.Dispose();
            });
        });

        WebViewList.Clear();

        return true;
    }


    private readonly HashSet<Signal<string>> _WebViewCloseSignal = new();
    private event Signal<string> _OnWebViewClose
    {
        add { if (value != null) lock (_WebViewCloseSignal) { _WebViewCloseSignal.Add(value); } }
        remove { lock (_WebViewCloseSignal) { _WebViewCloseSignal.Remove(value); } }
    }
    private readonly HashSet<Signal<string>> _WebViewOpenSignal = new();
    private event Signal<string> _OnWebViewOpen
    {
        add { if (value != null) lock (_WebViewOpenSignal) { _WebViewOpenSignal.Add(value); } }
        remove { lock (_WebViewOpenSignal) { _WebViewOpenSignal.Remove(value); } }
    }

}

File: android/app/src/main/java/info/bagen/dwebbrowser/microService/browser/mwebview/MultiWebViewController.kt
package info.bagen.dwebbrowser.microService.browser.mwebview

import android.annotation.SuppressLint
import androidx.appcompat.view.ContextThemeWrapper
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Stable
import androidx.compose.ui.platform.LocalDensity
import com.google.accompanist.web.WebContent
import com.google.accompanist.web.WebViewNavigator
import com.google.accompanist.web.WebViewState
import info.bagen.dwebbrowser.R
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import org.dweb_browser.browserUI.download.DownLoadObserver
import org.dweb_browser.dwebview.DWebView
import org.dweb_browser.dwebview.base.ViewItem
import org.dweb_browser.helper.Callback
import org.dweb_browser.helper.ChangeableList
import org.dweb_browser.helper.Signal
import org.dweb_browser.helper.SimpleSignal
import org.dweb_browser.helper.runBlockingCatching
import org.dweb_browser.helper.withMainContext
import org.dweb_browser.microservice.core.MicroModule
import org.dweb_browser.microservice.ipc.Ipc
import org.dweb_browser.microservice.ipc.helper.IpcEvent
import org.dweb_browser.window.core.WindowController
import org.dweb_browser.window.core.createWindowAdapterManager
import java.util.concurrent.atomic.AtomicInteger

/**
 * MWebView 是为其它模块提供 GUI 的程序模块，所以这里需要传入两个模块：localeMM 与 remoteMM
 * remoteMM 只是一层抽象，未来如何需要可以通过网络成实现
 */
@Stable
class MultiWebViewController(
  /**
   * 窗口控制器
   */
  val win: WindowController,
  /**
   * 控制者的通讯通道
   */
  val ipc: Ipc,
  /// 以下这两参数是用来构建DWebView的时候使用的
  private val localeMM: MicroModule,
  private val remoteMM: MicroModule,
) {
  companion object {
    private var webviewId_acc = AtomicInteger(1)
  }

  val webViewList = ChangeableList<MultiViewItem>()

  init {
    webViewList.onChange {
      updateStateHook()
    }
    val wid = win.id
    /// 提供渲染适配
    createWindowAdapterManager.renderProviders[wid] =
      @Composable { modifier ->
        val webViewScale = (LocalDensity.current.density * scale * 100).toInt()
        Render(modifier, webViewScale)
      }
    /// ipc 断开的时候，强制关闭窗口
    val off = ipc.onClose {
      win.close(force = true)
    }
    /// 窗口销毁的时候
    win.onClose {
      off();
      // 移除渲染适配器
      createWindowAdapterManager.renderProviders.remove(wid)
      // 清除释放所有的 webview
      for (item in webViewList) {
        closeWebView(item.webviewId)
      }
    }
  }

  fun isLastView(viewItem: MultiViewItem) = webViewList.lastOrNull() == viewItem
  fun isFistView(viewItem: MultiViewItem) = webViewList.firstOrNull() == viewItem
  val lastViewOrNull get() = webViewList.lastOrNull()
  fun getWebView(webviewId: String) = webViewList.find { it.webviewId == webviewId }

  data class MultiViewItem(
    override val webviewId: String,
    override val webView: DWebView,
    override val state: WebViewState,
    override val navigator: WebViewNavigator,
    override val coroutineScope: CoroutineScope,
    override var hidden: Boolean = false,
    val win: WindowController
  ) : ViewItem {
    internal val onReady = SimpleSignal()
    val onReadyListener = onReady.toListener()
  }

  var downLoadObserver: DownLoadObserver? = null

  /**
   * 打开WebView
   */
  suspend fun openWebView(url: String) = appendWebViewAsItem(createDwebView(url))

  suspend fun createDwebView(url: String): DWebView = withMainContext {
    val currentActivity = win.viewController.activity;// App.appContext
    val dWebView = DWebView(
      ContextThemeWrapper(currentActivity, R.style.Theme_dwebbrowser), remoteMM, DWebView.Options(
        url = url,
        /// 我们会完全控制页面将如何离开，所以这里兜底默认为留在页面
        onDetachedFromWindowStrategy = DWebView.Options.DetachedFromWindowStrategy.Ignore,
      ), currentActivity
    )
    dWebView
  }

  @SuppressLint("ClickableViewAccessibility")
  @Synchronized
  fun appendWebViewAsItem(dWebView: DWebView) = runBlockingCatching {
    withMainContext {
      val webviewId = "#w${webviewId_acc.getAndAdd(1)}"
      val state = WebViewState(WebContent.Url(dWebView.url ?: ""))
      val coroutineScope = CoroutineScope(CoroutineName(webviewId))
      val navigator = WebViewNavigator(coroutineScope)
      MultiViewItem(
        webviewId = webviewId,
        webView = dWebView,
        state = state,
        coroutineScope = coroutineScope,
        navigator = navigator,
        win = win,
      ).also { viewItem ->
        viewItem.webView.onReady { // 为了防止在窗口状态下，webview返回时失真问题。所以在webview加载完成后出发刷新
          viewItem.coroutineScope.launch { viewItem.onReady.emit() }
        }
        webViewList.add(viewItem)
        dWebView.onCloseWindow {
          closeWebView(webviewId)
        }
        webViewOpenSignal.emit(webviewId)
      }
    }
  }.getOrThrow()

  /**
   * 关闭WebView
   */
  suspend fun closeWebView(webViewId: String) =
    webViewList.find { it.webviewId == webViewId }?.let { viewItem ->
      webViewList.remove(viewItem)
      withContext(Dispatchers.Main) {
        viewItem.webView.destroy()
      }
      webViewCloseSignal.emit(webViewId)
      return true
    } ?: false

  /**
   * 移除所有列表
   */
  suspend fun destroyWebView(): Boolean {
    withContext(Dispatchers.Main) {
      webViewList.forEach { viewItem ->
        viewItem.webView.destroy()
      }
    }
    webViewList.clear()
    this.downLoadObserver?.close() // 移除下载状态监听
    return true
  }

  /**
   * 将指定WebView移动到顶部显示
   */
  suspend fun moveToTopWebView(webviewId: String): Boolean {
    val viewItem = webViewList.find { it.webviewId == webviewId } ?: return false
    webViewList.remove(viewItem)
    webViewList.add(viewItem)
    return true
  }

  fun getState(): JsonObject {
    val views = mutableMapOf<String, JsonElement>()
    debugMultiWebView("updateStateHook =>", webViewList.size)
    webViewList.forEachIndexed { index, it ->
      val viewItem = mutableMapOf<String, JsonElement>()
      viewItem["index"] = JsonPrimitive(index)
      viewItem["webviewId"] = JsonPrimitive(it.webviewId)
      viewItem["isActivated"] = JsonPrimitive(it.hidden)
      viewItem["mmid"] = JsonPrimitive(ipc.remote.mmid)
      viewItem["url"] = JsonPrimitive((it.state.content as WebContent.Url).url)
      views[it.webviewId] = JsonObject(viewItem)
    }
    val state = mutableMapOf<String, JsonElement>()
    state["wid"] = JsonPrimitive(win.id)
    state["views"] = JsonObject(views)
    return JsonObject(state)
  }

  private suspend fun updateStateHook() {
    ipc.postMessage(IpcEvent.fromUtf8("state", Json.encodeToString(getState())))
  }

  private val webViewCloseSignal = Signal<String>()
  private val webViewOpenSignal = Signal<String>()

  fun onWebViewClose(cb: Callback<String>) = webViewCloseSignal.listen(cb)
  fun onWebViewOpen(cb: Callback<String>) = webViewOpenSignal.listen(cb)


}